privacyidea/lib/tokens/radiustoken.py

```diff
@@ -168,24 +338,91 @@ def split_pin_pass(self, passw, user=None, options=None):
 
         return res, pin, otpval
 
+    @log_with(log)
+    @check_token_locked
+    def authenticate(self, passw, user=None, options=None):
+        """
+        do the authentication on base of password / otp and user and
+        options, the request parameters.
+
+        This is only called after it is verified, that the upper level is no challenge-request
+        or challenge-response
+
+        Here we contact the RADIUS server to validate the OtpVal.
+
+        :param passw: the password / otp
+        :param user: the requesting user
+        :param options: the additional request parameters
+
+        :return: tuple of (success, otp_count - 0 or -1, reply)
+
+        """
+        options = options or {}
+        res = False
+        radius_response = AccessReject
+        otp_counter = -1
+        reply = {'message': 'remote side denied access'}
+        otpval = passw
+
+        # should we check the pin locally?
+        if self.check_pin_local:
+            (_res, pin, otpval) = self.split_pin_pass(passw, user,
+                                                      options=options)
+
+            if not self.check_pin(pin):
+                return False, -1, {'message': "Wrong PIN"}
+
+        # attempt to retrieve saved state/result
+        state = options.get('radius_state')
+        result = options.get('radius_result')
+        if result is None:
+            radius_response = self._check_radius(otpval, options=options, radius_state=state)
+        else:
+            radius_response = result
+
+        if radius_response == AccessAccept:
+            res = True
+            reply = {'message': 'matching 1 tokens',
+                     'serial': self.get_serial(),
+                     'type': self.get_tokentype()}
+            otp_counter = 1
+
+        return res, otp_counter, reply
+
     @log_with(log)
     @check_token_locked
     def check_otp(self, otpval, counter=None, window=None, options=None):
+        """
```

Comment:
Is this function used in our production code? If I see correctly, it is currently only used in our RADIUS token tests?

Reply:
probably yes. But it looks fine, right?

Reply:
OK, let's keep it